observer/mon_conf.go

```diff
@@ -29,33 +30,38 @@ func (c *MonConf) validatePeriod() error {
 // value of the `Settings` field back to bytes, then passing it to the
 // `UnmarshalSettings` method of the `Configurer` type specified by the
 // `Kind` field.
-func (c MonConf) unmarshalConfigurer() (probers.Configurer, error) {
+func (c MonConf) unmarshalConfigurer() (probers.Configurer, string, error) {
 	kind := strings.Trim(strings.ToLower(c.Kind), " ")
 	configurer, err := probers.GetConfigurer(kind)
 	if err != nil {
-		return nil, err
+		return nil, kind, err
 	}
 	settings, _ := yaml.Marshal(c.Settings)
 	configurer, err = configurer.UnmarshalSettings(settings)
 	if err != nil {
-		return nil, err
+		return nil, kind, err
 	}
-	return configurer, nil
+	return configurer, kind, nil
 }
 
 // makeMonitor constructs a `monitor` object from the contents of the
 // bound `MonConf`. If the `MonConf` cannot be validated, an error
 // appropriate for end-user consumption is returned instead.
-func (c MonConf) makeMonitor() (*monitor, error) {
+func (c MonConf) makeMonitor(collectors map[string]map[string]*prometheus.Collector) (*monitor, error) {
 	err := c.validatePeriod()
 	if err != nil {
 		return nil, err
 	}
-	probeConf, err := c.unmarshalConfigurer()
+	probeConf, kind, err := c.unmarshalConfigurer()
 	if err != nil {
 		return nil, err
 	}
-	prober, err := probeConf.MakeProber()
+	pColls, exist := collectors[kind]
+	if !exist {
+		pColls = probeConf.Instrument()
+		collectors[kind] = pColls
```

Comment:
Modifying the input argument in this way is a valid way to solve this problem, but its dangerous for two reasons:
1) It's not very "idiomatic" go: Go generally tries very hard not to modify arguments, and when it does, it tries to make it very explicitly clear that it is doing so by taking _pointers to_ the objects that it is modifying. This makes this argument-modification surprising, and easily missed by someone quickly scanning the code.
2) It's not thread-safe if a future refactoring of this project initializes all of the monitors/probers in parallel in a bunch of goroutines.

So instead, I'd suggest having `makeMonitor` simply return the `kind` of monitor that it just created. Then `makeMonitors` can simply keep track of all of the different `kind`s it has seen. And at the end of its loop (where today you iterate over `range collectors`), it can iterate over `probers.Registry` and call the `Instrument()` method of all registered configurers whose `kind` it saw.